lib/powerpoint/slide/content.rb

```diff
@@ -0,0 +1,224 @@
+require 'erb'
+
+module Powerpoint
+  module Slide
+
+    # a modular slide with a title and any number of other elements, which are
+    # each individually rendered and added to the slide xml
+    class DefaultSlide
+      include Powerpoint::Util
+
+      attr_reader :elements, :title, :slide_layout, :content_xml, :images
+
+      def initialize(options={})
+        require_arguments [:elements, :title, :slide_layout], options
+        options.each {|k, v| instance_variable_set("@#{k}", v)}
+        @images = elements.select { |image| image.respond_to?(:embed_rid) }
+
+        @images.each_with_index do |image, i|
+          image.embed_rid = "rId#{i+2}"
+        end
+
+        @elements.each_with_index do |elem, i|
+          elem.id = 10 + i
+        end
+
+        @content_xml = gen_xml
+      end
+
+      def file_types
+        types = elements.map do |element|
+          element.file_type if element.respond_to?(:file_type)
+        end
+        types.compact.uniq
+      end
+
+      def gen_xml
+        xmls = elements.map { |e| e.render }
+        xmls.join('')
+      end
+
+      def render_rels
+        render_str('default_slide_rel.xml.erb')
+      end
+
+      def render_slide
+        render_str('default_slide.xml.erb')
+      end
+
+      def save(extract_path, index)
+        images.each do |image|
+          copy_media(extract_path, image.image_path)
+        end
+
+        File.open("#{extract_path}/ppt/slides/_rels/slide#{index}.xml.rels", 'w') do |f|
+          f << render_rels
+        end
+
+        File.open("#{extract_path}/ppt/slides/slide#{index}.xml", 'w') do |f|
+          f << render_slide
+        end
+
+      end
+    end
+
+    # renders an image for a PPT slide, snapping it to coordinates if passed.
+    class ImageContent
+      include Powerpoint::Util
+
+      attr_reader :image_name, :coords, :image_path, :id, :idx, :embed_rid
+      attr_writer :id, :idx, :embed_rid
+
+      def initialize(options={})
+        require_arguments [:image_path], options
+        options.each {|k, v| instance_variable_set("@#{k}", v)}
+        @idx = -1 unless @idx
+        @centered = false unless @centered
+        if !@coords
+          @coords = {}
+        else
+          @coords = boxed_coordinates(@coords, FastImage.size(@image_path), @centered)
+        end
+        @image_name = File.basename(@image_path)
+      end
+
+      def file_type
+        File.extname(image_name).gsub('.', '')
+      end
+
+      def render
+        render_str('picture.xml.erb')
+      end
+    end
+
+    class TableContent
+      include Powerpoint::Util
+
+      attr_reader :content, :id, :idx
+      attr_writer :id, :idx
+
+      def initialize(options={})
+        require_arguments [:content, :idx], options
+        options.each {|k, v| instance_variable_set("@#{k}", v)}
```

Comment:
Didn't take long for you to embrace the meta. Everyone loves their own reflection ;-)